exporter/sumologicexporter/sender.go

```diff
@@ -0,0 +1,257 @@
+		if sent {
+			currentTimeSeries = 0
+		}
+
+		// If log has been appended to body, increment the currentTimeSeries
+		if appended {
+			currentTimeSeries++
+		}
+	}
+
+	err := s.send(LogsPipeline, strings.NewReader(body.String()), fields)
+	if err != nil {
+		errs = append(errs, err)
+		droppedTimeSeries += currentTimeSeries
+	}
+
+	if len(errs) > 0 {
+		return droppedTimeSeries, componenterror.CombineErrors(errs)
+	}
+	return droppedTimeSeries, nil
+}
+
+func (s *sender) sendLogsJSONFormat(fields FieldsType) (int, error) {
+	var (
+		body              strings.Builder
+		errs              []error
+		droppedTimeSeries int
+		currentTimeSeries int
+	)
+
+	for _, record := range s.buffer {
+		data := s.filter.filterOut(record.Attributes())
+		data[logKey] = record.Body().StringVal()
+
+		nextLine, err := json.Marshal(data)
+		if err != nil {
+			droppedTimeSeries++
+			errs = append(errs, err)
+			continue
+		}
+
+		sent, appended, err := s.appendAndSend(bytes.NewBuffer(nextLine).String(), LogsPipeline, &body, fields)
+		if err != nil {
+			errs = append(errs, err)
+			if sent {
+				droppedTimeSeries += currentTimeSeries
+			}
+
+			if !appended {
+				droppedTimeSeries++
+			}
+		}
+
+		// If data was sent, cleanup the currentTimeSeries counter
+		if sent {
+			currentTimeSeries = 0
+		}
+
+		// If log has been appended to body, increment the currentTimeSeries
+		if appended {
+			currentTimeSeries++
+		}
+	}
+
+	err := s.send(LogsPipeline, strings.NewReader(body.String()), fields)
+	if err != nil {
+		errs = append(errs, err)
+		droppedTimeSeries += currentTimeSeries
+	}
+
+	if len(errs) > 0 {
+		return droppedTimeSeries, componenterror.CombineErrors(errs)
+	}
+	return droppedTimeSeries, nil
+}
+
+// appendAndSend appends line to the body and eventually sends data to avoid exceeding the request limit
+func (s *sender) appendAndSend(line string, pipeline PipelineType, body *strings.Builder, fields FieldsType) (bool, bool, error) {
+	var errors []error
+	// sent gives information if the data was sent or not
+	sent := false
+	// appended keeps state of appending new log line to the body
+	appended := true
+
+	if body.Len() > 0 && body.Len()+len(line) > s.config.MaxRequestBodySize {
+		err := s.send(LogsPipeline, strings.NewReader(body.String()), fields)
+		sent = true
+		if err != nil {
+			errors = append(errors, err)
+		}
+		body.Reset()
+	}
+
+	if body.Len() > 0 {
+		// Do not add newline if the body is empty
+		_, err := body.WriteString("\n")
+		if err != nil {
+			errors = append(errors, err)
+			appended = false
+		}
+	}
+
+	_, err := body.WriteString(line)
+	if err != nil && appended {
+		errors = append(errors, err)
+		appended = false
+	}
+
+	if len(errors) > 0 {
+		return sent, appended, componenterror.CombineErrors(errors)
+	}
+	return sent, appended, nil
+}
+
+// cleanBuffer zeroes buffer
+func (s *sender) cleanBuffer() {
+	s.buffer = (s.buffer)[:0]
+}
+
+// append adds log to the buffer
```

Comment:
No cheating in the comments :) explain that when buffer is full we send what has accumulated